Allow fine-tuning of MyST, Docutils and Sphinx settings

kdeldycke opened the pull request:
This PR:

- Adds a new `MYST_DOCUTILS_SETTINGS` and `MYST_SPHINX_SETTINGS` to configure each renderer and MyST parser
- Deprecates `MYST_EXTENSIONS`
- Feed `MYST_EXTENSIONS` to `MYST_DOCUTILS_SETTINGS` and `MYST_SPHINX_SETTINGS` in the mean time
- Forces initial header level to `<h2>` instead of `<h1>`, since Pelican will use the title field in the front-matter as the `<h1>` header of the page
- Adds typing to methods
- Cleans up the merge of users and default settings
- Documents the new configuration settings

Thanks to this PR, the behaviour of MyST parsing can be highly fine-tuned, which will invite users to more exploration.

kdeldycke committed:
Allow fine tuning of MyST, Docutils and Sphinx conf

kdeldycke commented:
All tests are passing and new features are documented. This PR is ready to be reviewed/merged upstream.

ashwinvis commented on README.md:
```diff
@@ -70,34 +72,86 @@ For more information on Pelican's default metadata format please visit the link
+DOCUTILS_SETTINGS = [
+    # Force the removal of comments from the document tree.
+    "strip_comments": True,
 ]
```
This should be a dictionary.

kdeldycke replied:
Ah yes, that's stupid. Fixed!

ashwinvis commented on pelican/plugins/myst_reader/_sphinx_renderer.py:
```diff
@@ -23,59 +27,60 @@ def get_div_body(html_output):
 
+    # Transfer MyST settings to Sphinx settings.
+    for param_id in myst_config.as_dict():
+        conf[f"myst_{param_id}"] = getattr(myst_config, param_id)
```
Can we iterate over key-value pairs directly using `myst_config.as_dict().items()`?

kdeldycke replied:
Good catch! Fixed!

ashwinvis commented on pelican/plugins/myst_reader/myst_reader.py:
```diff
@@ -50,20 +51,53 @@ class MySTReader(BaseReader):
+        for param_id in self.myst_config.as_dict():
+            self.docutils_settings[f"myst_{param_id}"] = getattr(
+                self.myst_config, param_id
+            )
```
- Why prefix the keys? We could follow the docs and expect users to set the keys as it is. With the prefix hard coded, it would be obvious and easier to copy paste from the docs.
- We can call `self.docutils_settings.update` at the end with Pelican's hardcoded settings so we ensure it is always there.

kdeldycke replied:
That part may not be clear enough I guess. Here is what I am trying to do.

The problem is that we use two parsers:
- The MySt parser to produce tokens
- The Docutils parser (or Sphinx) to produce HTML

These 2 steps are independents and each parser has its own way of getting parameters. The standalone MyST parser use the [same IDs as the frontmatter (local) configuration](https://myst-parser.readthedocs.io/en/latest/configuration.html#frontmatter-local-configuration), while the [Docutils parameters needs to be prefixed with `myst_`](https://myst-parser.readthedocs.io/en/latest/configuration.html#global-configuration) to be passed to its own MyST parser instance (like Sphinx).

To have the same behaviour accross the 2 components (+ Sphinx), I set the defaults of `self.docutils_settings` from `self.myst_config`, while prefixing them with `myst_`. Then I update the `self.docutils_settings` with the user's own values from `pelicanconf.py`.

That way a user can sets different configuration for the MySt parser and Docutils parser. But by default everything's aligned and there is no surprises.

This means a user can have in its `pelicanconf.py`:
```python
MYST_SETTINGS = {
    "gfm_only": True,
}

DOCUTILS_SETTINGS = {
    "myst_gfm_only": False,
}
```

kdeldycke replied:
Now with this explanation, how would you like to have the code change?

ashwinvis replied:
I studied your implementation a bit. I would like the user facing setttings to be like

```py
MYST_SPHINX_SETTINGS = {
    "myst_gfm_only": True,
}

MYST_DOCUTILS_SETTINGS = {
    "myst_gfm_only": False,
}
```
So they appear cohesive and the dictionary keys can easily be related to. Do you agree?


Also if possible add two unit tests which uses these variables. Right now the test cases work based on `MYST_EXTENSIONS` alone.

ashwinvis replied:
Alternatively it can also be

> ```python
> MYST_SPHINX_SETTINGS = {
>     "gfm_only": True,
> }
> 
> MYST_DOCUTILS_SETTINGS = {
>     "gfm_only": False,
> }
> ```

Btw, internally I don't think we should mix the two variables. I.e. we should keep the two rendering backends separately configurable. The user can easily do, `MYST_SPHINX_SETTINGS.update(MYST_DOCUTILS_SETTINGS)` in his/her pelicanconf.py.

kdeldycke replied:
OK!

So, to sum up the requested changes:
- I will add a new `MYST_SPHINX_SETTINGS` alongside `MYST_SETTINGS` and `DOCUTILS_SETTINGS`, that will be exclusively used to feed the Sphinx-based rendering mode
- I will update the documentation to invite users to use the short-form parameter IDs (i.e. `gfm_only` instead of `myst_gfm_only`)
- Make sure to translate the short parameter IDs to the `myst_` prefixed ones for Docutils and Sphinx

kdeldycke replied:
Thinking about it, and in light of your first observation, I'd rather not implement the prefixing with `myst_`. This might be too much magic and can add friction in the future if the sub-projects are changing their conventions.

kdeldycke replied:
Still, what do you think of auto-aligning the settings of the MyST parser and the Docutils parser? Should we get rid of this and invite the user to carefully setup its 3 constants `MYST_SETTINGS`, `MYST_SPHINX_SETTINGS` and `DOCUTILS_SETTINGS`?

I fear this might lower the user-friendliness of the plugin and increase support. Setting up a Pelican blog is still a lot of work, so if we can make this plugin just work it would increase its adoption.

ashwinvis replied:
No need for the ambiguous `MYST_SETTINGS`. Only MYST_SPHINX_SETTINGS and **MYST_DOCUTILS_SETTINGS**.

Sure in theory we could have delegated a single variable and do all the logic within the plugin. But I would rather have something maintainable and not prone to the changes upstream. Maybe in the future we can also implement a renderer for the Node JS based CLI for MyST. Which would require another variable...

ashwinvis replied:
In any case when user don't set these variables we can have reasonable defaults to reduce friction. You can choose the defaults! 🙂

kdeldycke replied:
OK I get it. And I didn't see that JS based CLI coming! 😅

The defaults are as they are now. I'll probably send another PR in a couple of weeks as I play around the iodiosyncracies of all the rendering stack.

Last question: should I get rid of `MYST_EXTENSIONS` too? In theory, nothing prevent the user to do:
```python
MYST_DOCUTILS_SETTINGS = {
    "myst_gfm_only": False,
    "myst_enable_extensions": [
        "amsmath",
        "tasklist",
    ],
}
```

kdeldycke replied:
OK I just pushed the last round of fixes.

ashwinvis commented on pelican/plugins/myst_reader/myst_reader.py:
```diff
@@ -190,27 +224,38 @@ def _extract_metadata(self, content):
     def _run_myst_to_tokens(self, content):
         """Execute the MyST parser and generate the syntax tree / tokens"""
-        return self.md_parser.parse(content)
+        md_parser = create_md_parser(self.myst_config, RendererHTML)
```
Move this to `__init__` as before so we don't instantiate a parser for each file.

kdeldycke replied:
Oh. I did not realized that method was called for each file. So yes, the original code was right to instantiate the parser in the `__init__`.

Sorry for the change. Just fixed it.

ashwinvis commented on pelican/plugins/myst_reader/myst_reader.py:
```diff
@@ -190,27 +224,38 @@ def _extract_metadata(self, content):
             self.force_sphinx
             or bib_files
-            or any(ext in ("dollarmath", "amsmath") for ext in self.myst_extensions)
+            or not self.myst_config.enable_extensions.isdisjoint(
```
```suggestion
            or self.myst_config.enable_extensions.intersection(
```

kdeldycke replied:
Indeed, that's more readable! Fixed!

ashwinvis commented on pelican/plugins/myst_reader/myst_reader.py:
```diff
@@ -190,27 +224,38 @@ def _extract_metadata(self, content):
-            return myst2html_with_docutils(content, self.myst_extensions, self.parser)
+            return myst2html_with_docutils(
+                content, self.docutils_settings, DocutilsParser()
+            )
```
Same as with `md_parser` before. Store parser as an attribute.

kdeldycke replied:
Yep, you are right again. Just reverted this change.

kdeldycke commented:
Thanks @ashwinvis for the detailed review of my PR! That was fast and on point!

I just fixed and addressed all your observations.

There is still one open code review to address but I need more feedback as I am not sure how you'd like the code to look.

kdeldycke commented:
I need to fix a couple of issue first before considering a last review.

ashwinvis commented:
The implementation overall looks quite solid. As for `MYST_EXTENSIONS` variable, we can deprecate it. If the user includes it in the pelicanconf.py file, we can broadcast a [`FutureWarning` ](https://docs.python.org/3/library/exceptions.html#FutureWarning) so they see it.

kdeldycke commented:
Everything has been fixed, everything is documented, and all tests are passing (but GitHub seems to be stuck right now).

We're ready for the final review.

ashwinvis committed:
Add test cases which use MYST_SPHINX_SETTINGS and MYST_DOCUTILS_SETTINGS

ashwinvis committed:
Fix parantheses in assert for standalone comment in test_comments

ashwinvis committed:
Try to interpret generic docutils exception

ashwinvis committed:
Readme gh actions badge and fix a typo

kdeldycke commented:
Thanks @ashwinvis for the merge upstream and the additional fixes! 💪
